src/Wind.cpp

```diff
@@ -0,0 +1,120 @@
+    if (mWindVelocitySubscriber)
+    {
+        mWindVelocitySubscriber->Unsubscribe();
+    }
+    mWindVelocitySubscriber = mNode->Subscribe("~/" + topicName, &Wind::readWindVelocity, this);
+    auto worldName = mModel->GetWorld()->Name();
+    gzmsg << "Wind: receiving wind commands from /gazebo/"
+          << worldName << "/" << topicName << endl;
+
+    mParameters = loadParameters(_sdf);
+}
+
+physics::LinkPtr Wind::getReferenceLink(physics::ModelPtr const _model, sdf::ElementPtr const _sdf) const
+{
+    if (_sdf->HasElement("link_name"))
+    {
+        physics::LinkPtr link = _model->GetLink(_sdf->Get<string>("link_name"));
+        if (!link)
+        {
+            string msg = "Wind: link " + _sdf->Get<string>("link_name") + " not found in model " + _model->GetName();
+            gzthrow(msg.c_str());
+        }
+        gzmsg << "Wind: reference link: " << link->GetName() << endl;
+        return link;
+    }
+    else if (_model->GetLinks().empty())
+    {
+        gzthrow("Wind: no link was defined in model!");
+    }
+    else
+    {
+        physics::LinkPtr link = _model->GetLinks().front();
+        gzmsg << "Wind: reference link not defined, using " << link->GetName() << " instead!" << endl;
+        return link;
+    }
+}
+
+Wind::EffectParameters Wind::loadParameters(sdf::ElementPtr el) const
+{
+    gzmsg << "Loading wind effect parameters" << endl;
+
+    EffectParameters parameters;
+    parameters.frontal_area = utilities::getParameter<double>("Wind", el, "frontal_area", "m2", 0);
+    parameters.lateral_area = utilities::getParameter<double>("Wind", el, "lateral_area", "m2", 0);
+    parameters.length_overall = utilities::getParameter<double>("Wind", el, "length_overall", "m", 0);
+    parameters.air_density = utilities::getParameter<double>("Wind", el, "air_density", "kg/m3", 1.12);
+    parameters.coefficients = utilities::getParameter<Vector3d>("Wind", el, "wind_coeffs",
+                                                                "", Vector3d(0, 0, 0));
+    return parameters;
+}
+
+void Wind::readWindVelocity(const ConstVector3dPtr &velocity)
+{
+    mWindVelocity = Vector3d(velocity->x(), velocity->y(), velocity->z());
+}
+
+Wind::Effects Wind::computeEffects()
+{
+    if (!mWindVelocity.X() && !mWindVelocity.Y())
+        return Effects{};
+
+    // Compute relative velocities and angle of attack
+    Angle model_yaw = mModel->WorldPose().Rot().Euler().Z();
+    // Wind direction in rad
+    double wind_direction_rad = atan2(-mWindVelocity.Y(), mWindVelocity.X());
+    Angle wind_direction(wind_direction_rad);
+    Angle wind_angle_from_bf(wind_direction - model_yaw); // Wind angle from body frame
+    Vector3d wind_velocity_bf(mWindVelocity.Length() * cos(wind_angle_from_bf.Radian()), mWindVelocity.Length() * sin(wind_angle_from_bf.Radian()), 0); // Wind velocity in the body frame
+    Vector3d relative_velocity = mModel->WorldLinearVel() - wind_velocity_bf;
```

Comment:
This combines the model velocity in the world frame with a wind velocity in the body frame. Axes aren't aligned, can't combine them.

To me, the math should look like (NOT REAL CODE):

```
// This is the wind velocity seen by the body, with X/Y axes aligned with the world
relative_wind_velocity_world = wind_velocity_world - model_velocity_world
// Now turned to align the X/Y axes with the body
world2body_orientation = inverse(body2world_orientation)
relative_wind_velocity_body = world2body_orientation * relative_wind_velocity_world
// We're only computing 2D effects
relative_wind_velocity_body.z() = 0;
```

where body2world_orientation is the orientation of the body in the world as a quaternion (i.e. the orientation you get from the pose of the body in the world).

This removes one angle calculation, which allows to test for zero-ness only on relative_wind_velocity_body, which is the only "wind" that matters (the only one that generates an effect).

Then, you should project this in 2D and test for zero-ness in a less strict manner (even though they are defined, things are usually ill-behaving when they are really small), i.e. instead of checking for equality to zero, I would do something like

```
if (relative_wind_velocity_body.norm() < 1e-3) {
   return Effects{};
}
```